### PR TITLE
Set 'Referer' HTTP request header

### LINKED ISSUE #2061: Wikimedia Maps HTTP 403 (acting like a bad bot)
[//]: # (Pick appropriate classifiers)
- Location:  Worker
- Schedule Name/ID: wikipedia*
- Task ID: https://farm.openzim.org/pipeline/94167b91-8b89-41de-a351-6287bab02c49

## Problem

The Wikipedia scrapers request map tiles from Wikimedia Foundation but have no permission.  This results in a lot of HTTP 403 responses.  It's possibly enough to be seen as a bad bot.

https://maps.wikimedia.org/img/osm-intl,5,40.415555555556,50.008611111111,260x260.png?lang=it&domain=it.wikipedia.org&title=Ateshgah_di_Baku&revid=137071110&groups=_4ddc15001962e3bddf2312007a0240d4b5316888
`Error: 403, Forbidden: Map tiles are restricted to Wikimedia & affiliated sites only. See https://wikitech.wikimedia.org/wiki/Maps/External_usage if you believe your usage supports the Movement. at Sat, 06 Jul 2024 18:55:57 GMT`

## Reproducing steps

Caused by scraping Wikipedia articles with maps.

### PR DESCRIPTION
**Pull request by audiodude**
Fixes #2061

Tested with Italian wikipedia download with article listed in bug.

### CODE REVIEW

**Comment by kelson42** [comment]
@audiodude Thank younvery much formthis simple but very important fix. One thing, why bot just taking the Mediwiki URL given as argument to MWoffliner in place of this fake "http://localhost/"?

**Comment by audiodude** [comment]
> @audiodude Thank you very much for this simple but very important fix. One thing, why bot just taking the Mediwiki URL given as argument to MWoffliner in place of this fake "http://localhost/"? To me this would more correct and probably more robust.

Because the WMF software looks for specific values of the `Referer` header. Presumably 'localhost' works because it is left in for local development. It seems better to put that, as a workaround, then `en.wikipedia.org`

**Comment by kelson42** [comment]
> > @audiodude Thank you very much for this simple but very important fix. One thing, why bot just taking the Mediwiki URL given as argument to MWoffliner in place of this fake "http://localhost/"? To me this would more correct and probably more robust.
> 
> Because the WMF software looks for specific values of the `Referer` header. Presumably 'localhost' works because it is left in for local development. It seems better to put that, as a workaround, then `en.wikipedia.org`

All WMF domains are part of the Regex look like, so why a hack like "localhost" cozld be better?

**Comment by audiodude** [comment]
> > > @audiodude Thank you very much for this simple but very important fix. One thing, why bot just taking the Mediwiki URL given as argument to MWoffliner in place of this fake "http://localhost/"? To me this would more correct and probably more robust.
> > 
> > 
> > Because the WMF software looks for specific values of the `Referer` header. Presumably 'localhost' works because it is left in for local development. It seems better to put that, as a workaround, then `en.wikipedia.org`
> 
> All WMF domains are part of the Regex look like, so why a hack like "localhost" could be better?

It just seems better to put a "dummy" value that is obviously a hack, than a WMF domain which is potentially misleading.

**Comment by kelson42** [comment]
> > > > @audiodude Thank you very much for this simple but very important fix. One thing, why bot just taking the Mediwiki URL given as argument to MWoffliner in place of this fake "http://localhost/"? To me this would more correct and probably more robust.
> 
> > > 
> 
> > > 
> 
> > > Because the WMF software looks for specific values of the `Referer` header. Presumably 'localhost' works because it is left in for local development. It seems better to put that, as a workaround, then `en.wikipedia.org`
> 
> > 
> 
> > All WMF domains are part of the Regex look like, so why a hack like "localhost" could be better?
> 
> 
> 
> It just seems better to put a "dummy" value that is obviously a hack, than a WMF domain which is potentially misleading.

Not really convinced, but I guessed you can argue that way. I believe we should have a small test validating this tomsecure next time we detect and can fix such kind of issue early.

**Comment by kelson42** [comment]
@audiodude Maybe we could just extend `saveArticles.test.ts` around the `London` test as https://en.wikipedia.org/wiki/London as a map?

**Comment by audiodude** [comment]
> @audiodude Maybe we could just extend `saveArticles.test.ts` around the `London` test as https://en.wikipedia.org/wiki/London as a map?

Added test in `downloader.test.ts` PTAL, thanks!

**Review by kelson42 — APPROVED** [review]
LGTM

File: src/Downloader.ts [hunk 1]
```diff
@@ -544,7 +546,9 @@ class Downloader {
           if (s3Resp?.Metadata?.etag) {
             this.arrayBufferRequestOptions.headers['If-None-Match'] = this.removeEtagWeakPrefix(s3Resp.Metadata.etag)
           }
-          const mwResp = await axios(url, this.arrayBufferRequestOptions)
+          // The 'Referer' header is set to get around WMF domain origin restrictions.
+          // See: https://github.com/openzim/mwoffliner/issues/2061
+          const mwResp = await axios(url, { ...this.arrayBufferRequestOptions, headers: { Referer: 'https://localhost/' } })
```
[the remark this answers is not in the document]
Comment (kelson42) [reply src/Downloader.ts:551]: @bd808 Was pretty much my proposal, see comments above and responses.

**Pull request merged**